core/metacling/src/TCling.cxx

```diff
@@ -1101,82 +1101,56 @@ static void LoadCoreModules(cling::Interpreter &interp)
-   // headers.
-   // FIXME: Use clang::ASTReader::makeModuleVisible.
-   std::stringstream declarations;
-   for (StringRef H : moduleHeaders) {
-      declarations << "#include \"" << H.str() << "\"\n";
-   }
-
-   // C99 decided that it's a very good idea to name a macro `I` (the letter I).
-   // This seems to screw up nearly all the template code out there as `I` is
-   // common template parameter name and iterator variable name.
-   // Let's follow the GCC recommendation and undefine `I` in case any of the
-   // core modules have defined it:
-   // https://www.gnu.org/software/libc/manual/html_node/Complex-Numbers.html
-   declarations << "#ifdef I\n #undef I\n #endif\n";
-
-   auto result = interp.declare(declarations.str());
+   cling::Transaction* T = nullptr;
+   interp.declare("/*This is decl is to get a valid sloc...*/;", &T);
+   SourceLocation ValidLoc = T->decls_begin()->m_DGR.getSingleDecl()->getLocStart();
+   clang::Sema &TheSema = CI.getSema();
+   // createImplicitModuleImportForErrorRecovery creates decls.
+   cling::Interpreter::PushTransactionRAII RAII(&interp);
+   std::vector<clang::Module*> TopLevelModulesToLoad;
+   for (auto I = moduleMap.module_begin(), E = moduleMap.module_end(); I != E; ++I) {
+     clang::Module *M = I->second;
+     assert(M);
+     if (M->IsSystem && !M->IsMissingRequirement) {
+       TheSema.createImplicitModuleImportForErrorRecovery(ValidLoc, M);
+       for (clang::Module *subModule : M->submodules())
+         TheSema.createImplicitModuleImportForErrorRecovery(ValidLoc, subModule);
+     }
+     else if (M->IsSystem && !M->IsMissingRequirement)
+       TopLevelModulesToLoad.push_back(M);
+   }
+   // Those two modules are required to be preloaded by ROOT. If they are not around we have a big problem.
+   clang::Module *CoreM = moduleMap.findModule("Core");
+   clang::Module *RIOM = moduleMap.findModule("RIO");
+   if (CoreM)
+      TheSema.createImplicitModuleImportForErrorRecovery(ValidLoc, CoreM);
+   else
+      Error("TCling::LoadCoreModules", "Internal error, couldn't find C++ module Core in modulemap.");
+   if (RIOM)
+      TheSema.createImplicitModuleImportForErrorRecovery(ValidLoc, RIOM);
+   else
+      Error("TCling::LoadCoreModules", "Internal error, couldn't find C++ module RIO in modulemap.");
 
-   if (result != cling::Interpreter::CompilationResult::kSuccess) {
-      Error("TCling::LoadCoreModules", "Couldn't parse core headers: %s\n", declarations.str().c_str());
+   // Loading all other modules in rootcling would mean injecting extra (needless) dependencies to the pcms.
+   bool fromRootCling = dlsym(RTLD_DEFAULT, "usedToIdentifyRootClingByDlSym");
+   if (fromRootCling)
+     return;
+
+   // FIXME: The performance of loading all modules at start up is comparable to the non-modules one. However we can
+   // do better: use the global module index to avoid loading all modules and load only the ones that contain the
+   // looked up identifier.
```

Comment:
I think we already asked Richard about that and he said it's only storing identifiers and therefor only useful in C (not C++).